File: 24.两两交换链表中的节点/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode swapPairs(ListNode head) {
        ListNode dummynode = new ListNode(0, head);
        if(head == null || head.next == null){
            return head;
        }
        ListNode pre = dummynode;
        ListNode cur = pre.next;
        ListNode tail = cur.next;
        while(tail != null){
            cur.next = tail.next;
            tail.next = cur;
            pre.next = tail;
            if(cur.next == null || tail.next == null){
                break;
            }
            else{
                pre = cur;
                cur = cur.next;
                tail = cur.next;
            }
        
        }
        return dummynode.next;
    }
}

// 递归版本
// class Solution {
//     public ListNode swapPairs(ListNode head) {
//         // base case 退出提交
//         if(head == null || head.next == null) return head;
//         // 获取当前节点的下一个节点
//         ListNode next = head.next;
//         // 进行递归
//         ListNode newNode = swapPairs(next.next);
//         // 这里进行交换
//         next.next = head;
//         head.next = newNode;

//         return next;
//     }
// } 

File: 24.两两交换链表中的节点/Solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        dummynode = ListNode()
        dummynode.next = head
        if head == None or head.next == None:
            return head
        pre = dummynode
        cur = pre.next
        tail = cur.next
        while tail:
            cur.next = tail.next
            tail.next = cur
            pre.next = tail
            if cur.next == None or tail.next == None:
                break
            else:
                pre = cur
                cur = cur.next
                tail = cur.next
        return dummynode.next
        